activerecord/lib/active_record/reflection.rb

```diff
@@ -199,10 +199,14 @@ def join_scope(table, foreign_table, foreign_klass)
 
         scope_chain_items.inject(klass_scope, &:merge!)
 
-        primary_key = join_primary_key
-        foreign_key = join_foreign_key
+        primary_key_column_names = Array(join_primary_key)
+        foreign_key_column_names = Array(join_foreign_key)
 
-        klass_scope.where!(table[primary_key].eq(foreign_table[foreign_key]))
+        primary_foreign_key_pairs = primary_key_column_names.zip(foreign_key_column_names)
+
+        primary_foreign_key_pairs.each do |primary_key_column_name, foreign_key_column_name|
+          klass_scope.where!(table[primary_key_column_name].eq(foreign_table[foreign_key_column_name]))
```

Comment:
There are couple of ways to "stack" the clauses. One way I have considered was prebuilding the `And` arel node manually and passing it to the ` klass_scope.where!` once. Like this:
```ruby
pk, fk = primary_foreign_key_pairs.pop
clause = table[pk].eq(foreign_table[fk])

primary_foreign_key_pairs.each do |primary_key_column_name, foreign_key_column_name|
  clause = clause.and(table[primary_key_column_name].eq(foreign_table[foreign_key_column_name]))
end

klass_scope.where!(clause)
```

To me the question here is basically how much we want to couple `Reflection` with arel internals. Even though it currently knows about the `.eq` semantics, I choose not to increase coupling by also using the `.and` node builder. Calling `.where!` feels like a higher-level abstraction which will handle the `And` node creation for us without us knowing about its existence. 

There is another approach I have considered but at this point it is unnecessary overcomplicated. We can introduce a new arel attribute type - `AttributesSet` which is a representation of several attributes. Then the code could have stayed the same as long as `table[primary_key].eq(foreign_table[foreign_key])` translates in

`AttributesSet.new([:column1, :column2]).eq(AttributesSet.new([:column3, :column4])`

I've elaborated a bit on the idea of the `AttributesSet` abstraction in a different PR - https://github.com/rails/rails/pull/47160 but in scope of this PR it is definitely an over-complication

Reply:
Using `.eq` is fine here because we already use it in other parts of Active Record internals. Arel is _for Rails_ internals, and technically the API isn't documented because it is internal. We shouldn't use it if we can use Active Record directly but in this case it's ok.